### ISSUE TITLE
run-pty seems to be leaking ANSI color escape sequences

### ISSUE THREAD

**Issue by Janiczek** [lydell/run-pty#28]
Here's something I saw when using run-pty with the Elm compiler.

<img width="755" alt="Screenshot 2022-09-14 at 13 24 54" src="https://user-images.githubusercontent.com/149425/190141361-ec59dd7e-73ba-47cb-8cdd-efd83d011d1c.png">

Please let me know what more details I can provide to help with this 🙂

**Comment by lydell** [comment]
My guess would be that this is caused by the injection of:

```
[ctrl+c] kill (pid 674)
[ctrl+z] dashboard
```

at the bottom, which is surprisingly tricky to do. So tricky that I’ve considered removing the whole feature from time to time, but I always come to the conclusion that I wouldn’t want to have colleagues ask me what keyboard combination to use to get back to the dashboard.

I’ve had bugs with that before, but now it’s been a while since I saw any of it. It has gotten pretty stable over the last couple of years.

Here are the main parts of it: https://github.com/lydell/run-pty/blob/c9030cd28d2c3a66b4b9524b665857b5440ab9f4/run-pty.js#L1391-L1469

Are you able to produce this consistently? Or pretty often? Or rarely? What is the full command you are running?

**Comment by Janiczek** [comment]
The full command is:
```
run-pty % yarn watch % ../dev/start staging --no-build --no-open % ./watch-compile.sh % yarn review --watch % yarn spec --parallel --watch % yarn test --watch
```
Would it be easier if the injection of keyboard shortcuts was at the top instead?

The issue happens for me consistently, AFAIK. My terminal is iTerm2 on macOS, `$TERM` is `xterm-256color` if that helps.

**Comment by lydell** [comment]
I also use iTerm2 on macOS. If you can figure out a small command that I can copy-paste and run to see it that would be great!

**Comment by Janiczek** [comment]
So, I've found this only happens after you kill one of the commands and then restart it:

https://user-images.githubusercontent.com/149425/190350738-4cb54630-ecc0-4ffd-8e12-d6ac53afe6e6.mp4

**Comment by Janiczek** [comment]
`yarn test --watch` feeds to `"test": "./test.sh"` which feeds to
```bash
#!/bin/sh
set -eu
cd $(dirname $0)
PATH="$PWD/node_modules/.bin:$PATH"

elm-test "$@"
```
with elm-test version `0.19.1-revision9`

`./watch-compile.sh` feeds into
```bash
#!/usr/bin/env bash

COLOR_OFF="\e[0m";
DIM="\e[2m";

function compile {
  find src -type f -name '*.elm' | xargs elm make --output=/dev/null
}

function run {
  clear;
  tput reset;
  echo -en "\033c\033[3J";

  echo -en "${DIM}";
  date -R;
  echo -en "${COLOR_OFF}";

  compile;
}

run;

chokidar src | while read WHATEVER; do
  run;
done;
```

**Comment by lydell** [comment]
Minimal, consistent reproduction:

1. Run `run-pty % elm-test --watch`
2. Cause a type error in Elm
3. Press `ctrl+c` and then `enter`

Now there are leaked color codes.

Great! Now I should be able to debug it.

**Comment by lydell** [comment]
I will probably merge and release #29 tomorrow, but if you’d like to try it out immediately you could run:

```
curl https://raw.githubusercontent.com/lydell/run-pty/fix-split-color-codes/run-pty.js > node_modules/run-pty/run-pty.js
```

**Comment by Janiczek** [comment]
@lydell So far it works flawlessly!

**Comment by lydell** [comment]
Released in v4.0.1.
